File: backend/app/schemas/outcomes.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime, date
from uuid import UUID


APP_STATUSES = ("saved", "applied", "screening", "interview", "offer_received", "selected", "rejected", "withdrawn")


class ApplicationCreate(BaseModel):
    hiring_requirement_id: UUID


class StatusTransitionIn(BaseModel):
    to_status: str = Field(
        ..., pattern="^(saved|applied|screening|interview|offer_received|selected|rejected|withdrawn)$"
    )
    note: Optional[str] = Field(None, max_length=1000)


class ApplicationEventOut(BaseModel):
    id: UUID
    application_id: UUID
    from_status: Optional[str] = None
    to_status: str
    actor: Optional[str] = None
    note: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ApplicationOut(BaseModel):
    id: UUID
    student_id: str
    hiring_requirement_id: UUID
    employer_id: Optional[UUID] = None
    status: str
    outcome: Optional[str] = None
    outcome_decided_at: Optional[datetime] = None
    applied_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ApplicationDetailOut(ApplicationOut):
    events: List[ApplicationEventOut] = []


class SkillFeedbackIn(BaseModel):
    skill_id: UUID
    expected_level: Optional[float] = Field(None, ge=0, le=1)
    observed_level: Optional[float] = Field(None, ge=0, le=1)
    comment: Optional[str] = Field(None, max_length=1000)


class EmployerFeedbackCreate(BaseModel):
    technical_ability: Optional[int] = Field(None, ge=1, le=5)
    communication: Optional[int] = Field(None, ge=1, le=5)
    problem_solving: Optional[int] = Field(None, ge=1, le=5)
    project_readiness: Optional[int] = Field(None, ge=1, le=5)
    role_readiness: Optional[int] = Field(None, ge=1, le=5)
    overall_rating: Optional[int] = Field(None, ge=1, le=5)
    interview_summary: Optional[str] = None
    overall_comment: Optional[str] = None
    skills: List[SkillFeedbackIn] = Field(default_factory=list, max_length=50)


class SkillFeedbackOut(BaseModel):
    id: UUID
    employer_feedback_id: UUID
    skill_id: UUID
    expected_level: Optional[float] = None
    observed_level: Optional[float] = None
    skill_gap: Optional[float] = None
    comment: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class EmployerFeedbackOut(BaseModel):
    id: UUID
    employer_id: UUID
    application_id: UUID
    student_id: str
    technical_ability: Optional[int] = None
    communication: Optional[int] = None
    problem_solving: Optional[int] = None
    project_readiness: Optional[int] = None
    role_readiness: Optional[int] = None
    overall_rating: Optional[int] = None
    interview_summary: Optional[str] = None
    overall_comment: Optional[str] = None
    status: str
    created_at: datetime
    updated_at: datetime
    skills: List[SkillFeedbackOut] = []

    class Config:
        from_attributes = True


class PlacementCreate(BaseModel):
    employer_id: UUID
    application_id: Optional[UUID] = None
    role_title: str = Field(..., min_length=2, max_length=200)
    location: Optional[str] = Field(None, max_length=200)
    joining_date: Optional[date] = None
    status: str = Field(..., pattern="^(offer_accepted|selected|joined|declined|not_joined)$")


class PlacementUpdate(BaseModel):
    role_title: Optional[str] = Field(None, min_length=2, max_length=200)
    location: Optional[str] = Field(None, max_length=200)
    joining_date: Optional[date] = None
    status: Optional[str] = Field(None, pattern="^(offer_accepted|selected|joined|declined|not_joined)$")
    verification_status: Optional[str] = Field(None, pattern="^(unverified|verified|disputed)$")


class PlacementOut(BaseModel):
    id: UUID
    student_id: str
    employer_id: UUID
    application_id: Optional[UUID] = None
    role_title: str
    location: Optional[str] = None
    joining_date: Optional[date] = None
    status: str
    outcome_source: str
    verification_status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AlignmentCreate(BaseModel):
    hiring_requirement_id: UUID
    course_name: str = Field(..., min_length=2, max_length=200)
    qualification_name: Optional[str] = Field(None, max_length=200)
    skill_id: Optional[UUID] = None
    coverage: Optional[float] = Field(None, ge=0, le=1)
    evidence_note: Optional[str] = None
    evidence_id: Optional[UUID] = None
    certification_id: Optional[UUID] = None


class AlignmentOut(BaseModel):
    id: UUID
    hiring_requirement_id: UUID
    course_name: str
    qualification_name: Optional[str] = None
    skill_id: Optional[UUID] = None
    coverage: Optional[float] = None
    evidence_note: Optional[str] = None
    evidence_id: Optional[UUID] = None
    certification_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DashboardOut(BaseModel):
    funnel: dict
    rates: dict
    denominators: dict
    top_required_skills: List[dict] = []
    top_observed_gaps: List[dict] = []
    unlinked_placements_n: int = 0
    window: dict


class IndustryOutcomeSignal(BaseModel):
    signal_version: str = "industry-outcome-signal-v1"
    role: Optional[str] = None
    location: Optional[str] = None
    skill: Optional[str] = None
    time_window: dict
    denominators: dict
    application_count: int
    interview_count: int
    selection_count: int
    placement_count: int
    feedback_count: int
    observed_demand: Optional[float] = None
    observed_skill_gap: Optional[float] = None
    min_n_met: bool

File: frontend/src/services/outcomes.ts
import { apiFetch } from "./api";

export type ApplicationStatus =
  | "saved"
  | "applied"
  | "screening"
  | "interview"
  | "offer_received"
  | "selected"
  | "rejected"
  | "withdrawn";

export type ApplicationEvent = {
  id: string;
  application_id: string;
  from_status: string | null;
  to_status: string;
  actor: string | null;
  note: string | null;
  created_at: string;
};

export type CandidateProfile = {
  full_name: string;
  college?: string | null;
  degree?: string | null;
  branch?: string | null;
  graduation_year?: number | null;
};

export type CandidateSkillMatch = {
  skill_id: string;
  skill_name?: string | null;
  importance?: string | null;
  required_level?: number | null;
  observed_level?: number | null;
  status?: "met" | "gap" | "unassessed" | string | null;
};

export type CandidateEvidenceItem = {
  id: string;
  type: string;
  title?: string | null;
  url?: string | null;
  description?: string | null;
};

export type Application = {
  id: string;
  student_id: string;
  hiring_requirement_id: string;
  employer_id?: string | null;
  status: ApplicationStatus;
  outcome: string | null;
  outcome_decided_at: string | null;
  applied_at: string | null;
  created_at: string;
  updated_at: string;
  requirement_title?: string | null;
  employer_name?: string | null;
  candidate_name?: string | null;
};

export type ApplicationDetail = Application & {
  events: ApplicationEvent[];
  candidate_profile?: CandidateProfile | null;
  candidate_skills?: CandidateSkillMatch[];
  candidate_evidence?: CandidateEvidenceItem[];
};


export type SkillFeedback = {
  id: string;
  employer_feedback_id: string;
  skill_id: string;
  skill_name?: string | null;
  expected_level: number | null;
  observed_level: number | null;
  skill_gap: number | null;
  comment: string | null;
  created_at: string;
};

export type EmployerFeedback = {
  id: string;
  employer_id: string;
  application_id: string;
  student_id: string;
  technical_ability: number | null;
  communication: number | null;
  problem_solving: number | null;
  project_readiness: number | null;
  role_readiness: number | null;
  overall_rating: number | null;
  interview_summary: string | null;
  overall_comment: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  skills: SkillFeedback[];
};

export type Placement = {
  id: string;
  student_id: string;
  employer_id: string;
  application_id: string | null;
  role_title: string;
  location: string | null;
  joining_date: string | null;
  status: string;
  outcome_source: string;
  verification_status: string;
  created_at: string;
  updated_at: string;
  employer_name?: string | null;
  requirement_title?: string | null;
};

export type Alignment = {
  id: string;
  hiring_requirement_id: string;
  course_name: string;
  qualification_name: string | null;
  skill_id: string | null;
  coverage: number | null;
  evidence_note: string | null;
  evidence_id: string | null;
  certification_id: string | null;
  created_at: string;
  updated_at: string;
};

export type Dashboard = {
  funnel: Record<string, number>;
  rates: {
    interview_rate: number;
    offer_rate: number;
    selection_rate: number;
    joining_rate_selection_base: number;
    joining_rate_application_base: number;
  };
  denominators: { applied_n: number; selected_n: number };
  top_required_skills: { skill_id: string; count: number }[];
  top_observed_gaps: { skill_id: string; avg_gap: number; feedback_count: number }[];
  unlinked_placements_n: number;
  window: { from: string | null; to: string | null };
};

export function createApplication(
  hiring_requirement_id: string,
  optionsOrStatus?: "saved" | "applied" | { initial_status?: "saved" | "applied"; note?: string },
  noteParam?: string
) {
  let initial_status: "saved" | "applied" = "applied";
  let note: string | undefined = undefined;
  if (typeof optionsOrStatus === "string") {
    initial_status = optionsOrStatus;
    note = noteParam;
  } else if (typeof optionsOrStatus === "object" && optionsOrStatus !== null) {
    if (optionsOrStatus.initial_status) initial_status = optionsOrStatus.initial_status;
    note = optionsOrStatus.note;
  }
  return apiFetch<Application>("/outcomes/applications", {
    method: "POST",
    body: JSON.stringify({ hiring_requirement_id, initial_status, note }),
  });
}

export type OpenRoleSkill = {
  skill_id: string;
  skill_name?: string | null;
  importance?: string | null;
  required_level?: number | null;
};

export type OpenRole = {
  id: string;
  title: string;
  role_key?: string | null;
  employer_id: string;
  employer_name?: string | null;
  location?: string | null;
  employment_type?: string | null;
  description?: string | null;
  created_at?: string | null;
  skills: OpenRoleSkill[];
  application?: { id: string; status: ApplicationStatus } | null;
};

export function listOpenRoles(search?: string) {
  return apiFetch<OpenRole[]>("/outcomes/open-roles", {
    params: search?.trim() ? { search: search.trim() } : undefined,
  });
}

export function listApplications(status?: string) {
  return apiFetch<Application[]>("/outcomes/applications", {
    params: status ? { status } : undefined,
  });
}

export function listApplicationsForEmployer(employerId: string) {
  return apiFetch<Application[]>(`/outcomes/employers/${employerId}/applications`);
}

export function listApplicationsForRequirement(requirementId: string) {
  return apiFetch<Application[]>(`/requirements/${requirementId}/applications`);
}

export function getApplication(id: string) {
  return apiFetch<ApplicationDetail>(`/outcomes/applications/${id}`);
}


export function transitionApplication(id: string, to_status: ApplicationStatus, note?: string) {
  return apiFetch<Application>(`/outcomes/applications/${id}/status`, {
    method: "PATCH",
    body: JSON.stringify({ to_status, note: note ?? null }),
  });
}

export function getFeedback(appId: string) {
  return apiFetch<EmployerFeedback>(`/outcomes/applications/${appId}/feedback`);
}

export function submitFeedback(
  appId: string,
  payload: {
    technical_ability?: number | null;
    communication?: number | null;
    problem_solving?: number | null;
    project_readiness?: number | null;
    role_readiness?: number | null;
    overall_rating?: number | null;
    interview_summary?: string | null;
    overall_comment?: string | null;
    skills?: {
      skill_id: string;
      expected_level?: number | null;
      observed_level?: number | null;
      comment?: string | null;
    }[];
  }
) {
  return apiFetch<EmployerFeedback>(`/outcomes/applications/${appId}/feedback`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function listPlacements() {
  return apiFetch<Placement[]>("/outcomes/placements");
}

export function listPlacementsForEmployer(employerId: string) {
  return apiFetch<Placement[]>(`/outcomes/employers/${employerId}/placements`);
}

export function getPlacement(id: string) {
  return apiFetch<Placement>(`/outcomes/placements/${id}`);
}

export function createPlacement(payload: {
  employer_id?: string | null;
  application_id?: string | null;
  role_title: string;
  location?: string | null;
  joining_date?: string | null;
  status?: string;
}) {
  return apiFetch<Placement>("/outcomes/placements", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updatePlacement(
  placementId: string,
  payload: {
    role_title?: string;
    location?: string | null;
    joining_date?: string | null;
    status?: string;
  }
) {
  return apiFetch<Placement>(`/outcomes/placements/${placementId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function confirmPlacement(placementId: string, verified: boolean = true) {
  return apiFetch<Placement>(`/outcomes/placements/${placementId}/confirm?verified=${verified}`, {
    method: "POST",
  });
}

export function listAlignments(hiring_requirement_id?: string) {
  return apiFetch<Alignment[]>("/outcomes/alignments", {
    params: hiring_requirement_id ? { hiring_requirement_id } : undefined,
  });
}

export function createAlignment(payload: {
  hiring_requirement_id: string;
  course_name: string;
  qualification_name?: string | null;
  skill_id?: string | null;
  coverage?: number | null;
  evidence_note?: string | null;
  evidence_id?: string | null;
  certification_id?: string | null;
}) {
  return apiFetch<Alignment>("/outcomes/alignments", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getDashboard(employer_id?: string) {
  return apiFetch<Dashboard>("/outcomes/dashboard", {
    params: employer_id ? { employer_id } : undefined,
  });
}
